pkg/vmprovider/providers/vsphere/vmprovider_vm_test.go

```diff
@@ -396,6 +396,87 @@ func vmTests() {
 				})
 			})
 
+			Context("VM Class Spec and ConfigSpec both contain GPU and DirectPath devices", func() {
+				BeforeEach(func() {
+					vmClass.Spec.Hardware.Devices = vmopv1alpha1.VirtualDevices{
+						VGPUDevices: []vmopv1alpha1.VGPUDevice{
+							{
+								ProfileName: "profile-from-class",
+							},
+						},
+						DynamicDirectPathIODevices: []vmopv1alpha1.DynamicDirectPathIODevice{
+							{
+								VendorID:    50,
+								DeviceID:    51,
+								CustomLabel: "label-from-class",
+							},
+						},
+					}
+
+					vmClass.Spec.Policies.Resources.Requests.Cpu = resource.MustParse("0")
+					vmClass.Spec.Policies.Resources.Limits.Cpu = resource.MustParse("0")
+
+					// Create the ConfigSpec with a GPU and a DDPIO device.
+					configSpec = &types.VirtualMachineConfigSpec{
+						Name: "dummy-VM",
+						DeviceChange: []types.BaseVirtualDeviceConfigSpec{
+							&types.VirtualDeviceConfigSpec{
+								Operation: types.VirtualDeviceConfigSpecOperationAdd,
+								Device: &types.VirtualPCIPassthrough{
+									VirtualDevice: types.VirtualDevice{
+										Backing: &types.VirtualPCIPassthroughVmiopBackingInfo{
+											Vgpu: "SampleProfile2",
+										},
+									},
+								},
+							},
+							&types.VirtualDeviceConfigSpec{
+								Operation: types.VirtualDeviceConfigSpecOperationAdd,
+								Device: &types.VirtualPCIPassthrough{
+									VirtualDevice: types.VirtualDevice{
+										Backing: &types.VirtualPCIPassthroughDynamicBackingInfo{
+											AllowedDevice: []types.VirtualPCIPassthroughAllowedDevice{
+												{
+													VendorId: 52,
+													DeviceId: 53,
+												},
+											},
+											CustomLabel: "SampleLabel2",
+										},
+									},
+								},
+							},
+						},
+					}
+				})
+
+				It("GPU and DirectPath devices from VM Class Spec.Devices are ignored", func() {
+					var o mo.VirtualMachine
+					Expect(vcVM.Properties(ctx, vcVM.Reference(), nil, &o)).To(Succeed())
+
+					devList := object.VirtualDeviceList(o.Config.Hardware.Device)
+					p := devList.SelectByType(&types.VirtualPCIPassthrough{})
+					Expect(p).To(HaveLen(2))
+
+					pciDev1 := p[0].GetVirtualDevice()
+					pciBacking1, ok1 := pciDev1.Backing.(*types.VirtualPCIPassthroughVmiopBackingInfo)
+					Expect(ok1).Should(BeTrue())
+					Expect(pciBacking1.Vgpu).To(Equal("SampleProfile2"))
+					Expect(pciBacking1.Vgpu).ToNot(Equal(vmClass.Spec.Hardware.Devices.VGPUDevices[0].ProfileName))
+
+					pciDev2 := p[1].GetVirtualDevice()
+					pciBacking2, ok2 := pciDev2.Backing.(*types.VirtualPCIPassthroughDynamicBackingInfo)
+					Expect(ok2).Should(BeTrue())
+					Expect(pciBacking2.AllowedDevice).To(HaveLen(1))
+					Expect(pciBacking2.AllowedDevice[0].VendorId).To(Equal(int32(52)))
+					Expect(pciBacking1.Vgpu).ToNot(Equal(vmClass.Spec.Hardware.Devices.DynamicDirectPathIODevices[0].VendorID))
```

Comment:
Are the `pciBacking1.Vgpu` assertions here a copy-pasta? This is already asserted above, and a little odd to be asserting the vGPU to not equals`vmClass...DynamicDirectPathIODevices[0]` fields.